third_party/amd/lib/TritonAMDGPUTransforms/InThreadTranspose.cpp

```diff
@@ -0,0 +1,620 @@
+            traverseCFForValueUses(candidate->getResult(forwardIdx));
+        if (failed(forwardSearch))
+          return failure();
+        foundNetwork.sharedMemVals.append(forwardSearch.value().transitiveCF);
+        nextTraversalStep.append(forwardSearch.value().ops);
+      }
+    }
+    traversalStep = std::move(nextTraversalStep);
+  }
+  return foundNetwork;
+}
+
+unsigned getDimRepeats(RankedTensorType type, int dimIdx) {
+  auto loadEnc = type.getEncoding();
+  auto blockedEnc = dyn_cast<ttg::BlockedEncodingAttr>(loadEnc);
+  if (!blockedEnc)
+    return 0;
+  auto sizePerThread = blockedEnc.getSizePerThread();
+  auto lanes = blockedEnc.getThreadsPerWarp();
+  auto warps = blockedEnc.getWarpsPerCTA();
+  auto shape = type.getShape();
+  int repeats =
+      shape[dimIdx] / (sizePerThread[dimIdx] * lanes[dimIdx] * warps[dimIdx]);
+  return std::max(1, repeats);
+}
+
+llvm::FailureOr<loadStoreLoadPatternComponents>
+matchThreadRakePattern(Value operand) {
+  auto opTensorTy = cast<RankedTensorType>(operand.getType());
+  auto opEnc = opTensorTy.getEncoding();
+  auto opDotOpEnc = dyn_cast<ttg::DotOperandEncodingAttr>(opEnc);
+  if (!opDotOpEnc)
+    return failure();
+
+  int kDimNum = opDotOpEnc.getOpIdx() == 0 ? 1 : 0;
+  // TODO: support wmma
+  if (!isa<ttg::AMDMfmaEncodingAttr, ttg::AMDWmmaEncodingAttr>(
+          opDotOpEnc.getParent())) {
+    LDBG("Operand's parent encoding is not MFMA");
+    return failure();
+  }
+
+  // Find nearest local_load
+  auto localLoadSearch = findAllDefiningOps<ttg::LocalLoadOp>(operand);
+  if (failed(localLoadSearch)) {
+    LDBG("Failed to traverse local loads");
+    return failure();
+  }
+
+  if (localLoadSearch.value().size() == 0) {
+    LDBG("Did not find local load operation");
+    return failure();
+  }
+
+  loadStoreLoadPatternComponents pattern;
+
+  SetVector<Operation *> visited;
+  for (auto lLoad : localLoadSearch.value()) {
+    // find local_alloc, local_store, local_load and ttg.memdesc_subview
+    // operations
+    auto sharedMemSearch = findReachableSMemOps(lLoad, visited);
+    if (failed(sharedMemSearch)) {
+      LDBG("Failed to traverse shared memmory operation network");
+      return failure();
+    }
+    pattern = sharedMemSearch.value();
+  }
+
+  if (pattern.localMemStores.empty()) {
+    LDBG("Did not find local alloc or store operations");
+    return failure();
+  }
+
+  for (auto localMemStore : pattern.localMemStores) {
+    LDBG("processing local mem store operation: " << *localMemStore);
+    // check if it is a local alloc with no predecessor
+    if (localMemStore->getNumOperands() == 0)
+      continue;
+    Value loadCandidate = localMemStore->getOperand(0);
+    auto loadedEnc =
+        cast<RankedTensorType>(loadCandidate.getType()).getEncoding();
+    auto blockedEnc = dyn_cast<ttg::BlockedEncodingAttr>(loadedEnc);
+    if (!blockedEnc)
+      return failure();
+    auto order = blockedEnc.getOrder();
+    if (order[0] == kDimNum) {
```

Comment:
Add comment to mention that we are meaning to handle non-K-continous cases in this pass.